### PR TITLE
persist: flush out update batches to s3 to bound memory use

### PR DESCRIPTION
**Pull request by danhhz**
Previously, a BatchBuilder would buffer all updates in memory and flush
them all out as a single BlobTraceBatchPart when the persist user called
BatchBuilder::finish. Now, whenever a BatchBuilder::add call causes
ColumnarRecordsVecBuilder to finish a chunk of updates, they are sent
off in a task to be encoded and uploaded to blob storage. If too many of
these are outstanding (perhaps s3 is down or slow), BatchBuilder::add is
backpressured until the number that are outstanding is back under the
threshold.

Touches #12893

### Motivation

  * This PR adds a known-desirable feature.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by danhhz** [comment]
FYI @aljoscha no need to review yet, I still need to add tests

File: src/persist-client/src/batch.rs
Comment (danhhz) [inline src/persist-client/src/batch.rs:257]: not to self, I think this needs to increment max_ts somehow to be a proper upper

Comment (aljoscha) [reply src/persist-client/src/batch.rs:257]: I think yes, the `upper` of a batch must not be `<=` any of the timestamps in the batch.

**Review by aljoscha — COMMENTED** [review]
Looks correct, but I'll wait till you added tests and tell me I should review it.

File: src/persist-client/src/batch.rs
Comment (aljoscha) [reply src/persist-client/src/batch.rs:257]: maybe parts don't have an upper? just spitballing...

Comment (danhhz) [reply src/persist-client/src/batch.rs:257]: hmm they don't technically need the description at all. it's just a thing we have that we think might be useful for _hand-wave_ future debugging

**Comment by danhhz** [comment]
Okay this is ready to go! I also manually verified that this is doing the right thing on the open-loop benchmark and that our max throughput is unaffected. We're still limited by fake data generation, but it actually looks like this _improved_ (!) our max throughput because we've now parallelized encoding with adding data to the BatchBuilder.

image

**Comment by aljoscha** [comment]
> We're still limited by fake data generation, but it actually looks like this _improved_ (!) our max throughput because we've now parallelized encoding with adding data to the BatchBuilder.

\o/

File: src/persist-client/src/batch.rs
Comment (aljoscha) [inline src/persist-client/src/batch.rs:192]: Before I started the review and saw this I was worried that we might now be producing too many small blobs! :raised_hands: 

I had started thinking that we might want to expose multi-part uploads from the `Blob` API, that way, we could thread pipelining all the way through to `Blob`. But it's not as straightforward as just tweezing apart `Blob::set()`.

File: src/persist-client/src/batch.rs
Comment (aljoscha) [inline src/persist-client/src/batch.rs:225]: :+1:

File: src/persist-client/src/batch.rs
Comment (aljoscha) [inline src/persist-client/src/batch.rs:276]: Yeah, I've dabbled with this _inclusive upper_ in my persistent source examples. But it can be confusing to people that are used to timely/dd uppers. No need to change anything here, btw!

Also, nice that you thought of the empty antichain as a valid upper to get around this problem for now!

File: src/persist-client/src/batch.rs
Comment (aljoscha) [inline src/persist-client/src/batch.rs:232]: Can just use `self.blob`, no need to clone.

File: src/persist-client/src/batch.rs
Comment (aljoscha) [inline src/persist-client/src/batch.rs:411]: I think I get your motivation for this: we only log when we have to wait on pending uploads. But this could also be an `#[instrument]`, on the method to keep it simpler. Feel free to keep as is, though. Maybe use `batch::finish_upload` as the name?

File: src/persist-client/src/batch.rs
Comment (aljoscha) [inline src/persist-client/src/batch.rs:467]: nit: "allowed"

File: src/persist-client/src/batch.rs
Comment (aljoscha) [inline src/persist-client/src/batch.rs:365]: I'm guessing you noticed this just from eyeballing, not any observations in the benchmark?

**Review by aljoscha — APPROVED** [review]
Very nice! I like that we're making progress on these low-hanging fruits, and cut down on `TODOs` in the code.

File: src/persist-client/src/batch.rs
Comment (danhhz) [reply src/persist-client/src/batch.rs]: done and done! you're right that I was over-thinking this

File: src/persist-client/src/batch.rs
Comment (danhhz) [reply src/persist-client/src/batch.rs]: done

**Review by danhhz — COMMENTED** [review]
TFTR!

**Pull request merged**
